openwisp_notifications/handlers.py

```diff
@@ -316,3 +317,19 @@ def schedule_object_notification_deletion(instance, created, **kwargs):
         tasks.delete_ignore_object_notification.apply_async(
             (instance.pk,), eta=instance.valid_till
         )
+
+
+def register_notification_cache_update(model, signal, signal_uid=None):
+    signal.connect(
+        update_notification_cache,
+        sender=model,
+        dispatch_uid='{}_{}_update_notification_cache'.format(
+            signal_uid, model.__name__
+        ),
+    )
+
+
+def update_notification_cache(sender, instance, **kwargs):
+    content_type = ContentType.objects.get_for_model(instance)
+    cache_key = Notification._cache_key(content_type.id, instance.id)
+    cache.delete(cache_key)
```

Comment:
here's another thing I didn't spot before, we should call these operations after previous changes have been committed to the database using `transaction.on_commit`, otherwise we risk having different behaviour in prod than what we have in development.

In production, there are many occasions in which sigals fire before changes have been committed to the DB, leading to bugs. We fixed most of these issues but we're still fixing others: https://github.com/openwisp/openwisp-notifications/pull/167

Therefore it's better to prevent this here.

Could be done like the following:

```suggestion
from django.db import transaction

def update_notification_cache(sender, instance, **kwargs):
    def invalidate_cache():
        content_type = ContentType.objects.get_for_model(instance)
        cache_key = Notification._cache_key(content_type.id, instance.id)
        cache.delete(cache_key)
    # execute cache invalidation only after changes have been committed to the DB
    transaction.on_commit(invalidate_cache)
```

Reply:
@nemesisdesign Thanks